src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/metering/MeteringProcessor.scala

```diff
@@ -0,0 +1,86 @@
+package tech.cryptonomic.nautilus.cloud.adapters.metering
+
+import java.time.Instant
+
+import cats.Monad
+import cats.implicits._
+import com.typesafe.scalalogging.StrictLogging
+import tech.cryptonomic.nautilus.cloud.domain.apiKey.{ApiKey, ApiKeyRepository, ApiKeyStats}
+import tech.cryptonomic.nautilus.cloud.domain.metering.api.MeteringApiRepository
+import tech.cryptonomic.nautilus.cloud.domain.metering.stats.{AggregatedMeteringStats, MeteringStatsRepository}
+import tech.cryptonomic.nautilus.cloud.domain.user.{User, UserRepository}
+
+import scala.language.higherKinds
+
+/** Gathers and processes metering data */
+class MeteringProcessor[F[_]: Monad](
+    meteringStatsRepository: MeteringStatsRepository[F],
+    meteringApiRepository: MeteringApiRepository[F],
+    apiKeyRepository: ApiKeyRepository[F],
+    userRepository: UserRepository[F],
+    meteringApiConfig: MeteringApiConfig
+) extends StrictLogging {
+
+  /** Process for getting current users, fetching stats, aggregating and inserting to the DB */
+  def process(): F[Unit] =
+    for {
+      users <- userRepository.getUsers()().map(_.result)
+      _ = logger.debug(s"Got users from DB: $users")
+      meteringStats <- meteringStatsRepository.getLastStats(users.map(_.userId))
+      _ = logger.debug(s"Got last stats from DB: $meteringStats")
+      validApiKeys <- users.map { user =>
+        val periodStart = meteringStats
+          .find(_.userId == user.userId)
+          .map(_.periodStart)
+          .getOrElse(Instant.ofEpochSecond(0))
+        apiKeyRepository
+          .getUserActiveKeysInGivenRange(user.userId, periodStart, Instant.now())
+      }.sequence
+      _ = logger.debug(s"Valid API keys: ${validApiKeys.flatten}")
+      apiKeyStats <- if (validApiKeys.flatten.isEmpty) {
+        List.empty[ApiKeyStats].pure[F]
+      } else {
+        fetchApiKeyStats(validApiKeys.flatten, meteringStats)
```

Comment:
no, it's conditional filtering, when the list is empty it would return results for all api keys, when there are api keys in the filter it will return only those

Reply:
😕 ... I see that if `validApiKeys` is empty `apiKeyStats` is empty... and `fetchApiKeyStats` won't be called... therefore I see two possible situations

1. is `fetchApiKeyStats` called otherwise somewhere else, and it will consider empty keys as a fetch-all query?
2. if not, what's the use of having `fetchApiKeyStats` return anything other then an empty stats for empty keys?

Which is the case, here?

Reply:
1. `fetchApiKeyStats` will consider empty keys as fetch-all query

Reply:
thanks